### ISSUE TITLE
Generated fact contains wrong data 

### ISSUE THREAD

**Issue by hferentschik** [jenkins-x/ext-jacoco#95]
Below you find an example _Fact_ as currently generated by the app. There seem to be several things not quite right:

* The fact id is '0'. In fact, the code never sets it, meaning all facts will have the same id. What is the intention here? Do we want a random id (aka uuid) in order to identify this particular fact at any time or should the id be some sort of hash over the data, so that two facts which are identical in data get the same id?
* The measurement type says "percent", but we are processing counters. Should the type be "count" or we want to calculate percentages?
* We are only processing the top level aggregated counters in these facts. The underlying jacoco report is more detailed. Is it intentional to only look at the aggregate
*  ' Instructions-Coverage', 'Complexity-Coverage', etc is imo misleading. It is not "coverage" which we record, but lines "covered". That's an issue in the _Fact_ definition though

```yaml
facts:
  - factType: jx.coverage
    id: 0
    measurements:
    - measurementType: percent
      measurementValue: 3
      name: Instructions-Coverage
    - measurementType: percent
      measurementValue: 5
      name: Instructions-Missed
    - measurementType: percent
      measurementValue: 8
      name: Instructions-Total
    - measurementType: percent
      measurementValue: 1
      name: Lines-Coverage
    - measurementType: percent
      measurementValue: 2
      name: Lines-Missed
    - measurementType: percent
      measurementValue: 3
      name: Lines-Total
    - measurementType: percent
      measurementValue: 1
      name: Complexity-Coverage
    - measurementType: percent
      measurementValue: 1
      name: Complexity-Missed
    - measurementType: percent
      measurementValue: 2
      name: Complexity-Total
    - measurementType: percent
      measurementValue: 1
      name: Methods-Coverage
    - measurementType: percent
      measurementValue: 1
      name: Methods-Missed
    - measurementType: percent
      measurementValue: 2
      name: Methods-Total
    - measurementType: percent
      measurementValue: 1
      name: Classes-Coverage
    - measurementType: percent
      measurementValue: 0
      name: Classes-Missed
    - measurementType: percent
      measurementValue: 1
      name: Classes-Total
    name: ""
    original:
      mimetype: application/xml
      tags:
      - jacoco.xml
      url: https://raw.githubusercontent.com/hf-bee/spring-boot-test/gh-pages/jenkins-x/jacoco/hf-bee/spring-boot-test/PR-5/1/target/site/jacoco/jacoco.xml
    statements: []
    tags:
    - jacoco
```

@deanesmith, @pmuir - WDYT?

**Comment by hferentschik** [comment]
As reference here is the XML report:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE report PUBLIC "-//JACOCO//DTD Report 1.1//EN" "report.dtd">
<report name="demo">
   <sessioninfo id="59511b2e-2df8-11e9-b92d-0a580a14090f-pod-fac118-ca1c2bdb" start="1549888077882" dump="1549888087137" />
   <package name="com/example/springboottest">
      <class name="com/example/springboottest/DemoApplication" sourcefilename="DemoApplication.java">
         <method name="&lt;init&gt;" desc="()V" line="7">
            <counter type="INSTRUCTION" missed="0" covered="3" />
            <counter type="LINE" missed="0" covered="1" />
            <counter type="COMPLEXITY" missed="0" covered="1" />
            <counter type="METHOD" missed="0" covered="1" />
         </method>
         <method name="main" desc="([Ljava/lang/String;)V" line="11">
            <counter type="INSTRUCTION" missed="5" covered="0" />
            <counter type="LINE" missed="2" covered="0" />
            <counter type="COMPLEXITY" missed="1" covered="0" />
            <counter type="METHOD" missed="1" covered="0" />
         </method>
         <counter type="INSTRUCTION" missed="5" covered="3" />
         <counter type="LINE" missed="2" covered="1" />
         <counter type="COMPLEXITY" missed="1" covered="1" />
         <counter type="METHOD" missed="1" covered="1" />
         <counter type="CLASS" missed="0" covered="1" />
      </class>
      <sourcefile name="DemoApplication.java">
         <line nr="7" mi="0" ci="3" mb="0" cb="0" />
         <line nr="11" mi="4" ci="0" mb="0" cb="0" />
         <line nr="12" mi="1" ci="0" mb="0" cb="0" />
         <counter type="INSTRUCTION" missed="5" covered="3" />
         <counter type="LINE" missed="2" covered="1" />
         <counter type="COMPLEXITY" missed="1" covered="1" />
         <counter type="METHOD" missed="1" covered="1" />
         <counter type="CLASS" missed="0" covered="1" />
      </sourcefile>
      <counter type="INSTRUCTION" missed="5" covered="3" />
      <counter type="LINE" missed="2" covered="1" />
      <counter type="COMPLEXITY" missed="1" covered="1" />
      <counter type="METHOD" missed="1" covered="1" />
      <counter type="CLASS" missed="0" covered="1" />
   </package>
   <counter type="INSTRUCTION" missed="5" covered="3" />
   <counter type="LINE" missed="2" covered="1" />
   <counter type="COMPLEXITY" missed="1" covered="1" />
   <counter type="METHOD" missed="1" covered="1" />
   <counter type="CLASS" missed="0" covered="1" />
</report>
```

**Comment by deanesmith** [comment]
My suggestions.  Need @pmuir  to weight in.
- A uuid for the fact ID could be helpful, though I am not yet fully understanding of exactly how
- The data type should not be percent, count seems right
- Top level data should be fine for now
- Can we not describe this better as a "fact"?  Your point is quite valid.

**Comment by hferentschik** [comment]
Turns out, we also don't set the name.

**Comment by jglick** [comment]
See also https://github.com/jenkins-x/jx/pull/3066.

**Comment by hferentschik** [comment]
> See also [jenkins-x/jx#3066](https://github.com/jenkins-x/jx/pull/3066).

Cheers - https://github.com/jenkins-x-apps/jx-app-jacoco/issues/99

**Comment by hferentschik** [comment]
So, for now, I go with:

* Generate unique name of the form jacoco-jx.coverage-<pipeline-name>
* ID will be removed in the future version of the Fact CRD
* I switch the measurement type to count (if in the feature we need percent we need to update this and do some calculations)

**Comment by hferentschik** [comment]
Merged via PR #100
